README.md
# Profile Card

A responsive profile card with a theme toggle feature that allows users to switch between light and dark themes. The profile card contains a profile picture, headline, description, personal links, and social links.

![Profile Card Preview](./preview.png)

## Usage as a GitHub Page

1. Fork this repository to your GitHub account.
2. Go to the repository settings.
3. Under the "GitHub Pages" section, select the "main" branch as the source and click "Save".
4. Your profile card will be published at `https://<your-username>.github.io/<repository-name>/`.

## Customization

1. Open the `config.js` file in the project.
2. Modify the `profileCardConfig` object to customize your profile card with your own data, links, and themes.

```javascript
const profileCardConfig = {
  profile: {
    picture: "http://placekitten.com/g/200/300",
    headline: "Kitten",
    description: "Your description goes here...Your description goes here...",
  },
  personalLinks: [
    { text: "Portfolio", url: "#" },
    { text: "Contact", url: "#" },
  ],
  socialLinks: [
    { icon: "logo-github", url: "#" },
    { icon: "logo-mastodon", url: "#" },
    { icon: "logo-twitter", url: "#" },
    { icon: "logo-instagram", url: "#" },
  ],
  themes: {
    light: {
      "--bg-color": "#fff",
      "--text-color": "#000",
      "--card-bg-color": "#e4e4e4",
      "--accent-color": "#007BFF",
    },
    dark: {
      "--bg-color": "#1a1a1a",
      "--text-color": "#fff",
      "--card-bg-color": "#333",
      "--accent-color": "#00BCD4",
    },
  },
};
```

3. Commit and push your changes to the repository.
4. Your profile card will be updated with your customizations.

## License

This project is open-source and available under the MIT License. See [LICENSE](LICENSE) for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.0.1] - 2022-05-18

### Added

- Added `config.js` for easier customization.
- Added README.md and CHANGELOG.md files.

## [1.0.0] - 2023-05-18

### Initial Release

- Profile card with theme toggle (light and dark mode).
- Profile picture, headline, and description.
- Personal and social links.


script.js
import { profileCardConfig } from "./config.js";

class ProfileCard {
    constructor(profile, personalLinks, socialLinks, themes) {
        this.profile = profile;
        this.personalLinks = personalLinks;
        this.socialLinks = socialLinks;
        this.themes = themes;
    }

    setTheme(name) {
        const theme = this.themes[name];
        for (let prop in theme) {
            document.documentElement.style.setProperty(prop, theme[prop]);
        }
    }

    populateProfile() {
        const profileElem = document.getElementById('profile');
        profileElem.innerHTML = `
            <img src="${this.profile.picture}" alt="Profile Picture" >
            <h2>${this.profile.headline}</h2>
            <p>${this.profile.description}</p>
        `;
    }

    populatePersonalLinks() {
        const personalLinksElem = document.getElementById('personal-links');
        personalLinksElem.innerHTML = this.personalLinks.map(link => `
            <a href="${link.url}" title="${link.text}">
                <button>${link.text}</button>
            </a>
        `).join("");
    }

    populateSocialLinks() {
        const socialLinksElem = document.getElementById('social-links');
        socialLinksElem.innerHTML = this.socialLinks.map(link => `
            <a href="${link.url}" title="${link.icon}">
                <ion-icon name="${link.icon}"></ion-icon>
            </a>
        `).join("");
    }


    initializeThemeToggle(moon, sun) {

        moon.addEventListener('click', (e) => {
            this.setTheme('dark');
            moon.style.display = 'none';
            sun.style.display = 'block';
        });

        sun.addEventListener('click', (e) => {
            this.setTheme('light');
            sun.style.display = 'none';
            moon.style.display = 'block';
        });
    }

    initializeCardTransform() {
        const card = document.getElementById('card');

        let timeoutID;
        card.addEventListener('mouseenter', function (e) {
            timeoutID = setTimeout(() => {
                card.style.transition = 'none';
            }, 0);
        });

        card.addEventListener('mousemove', function (e) {
            const rect = card.getBoundingClientRect();
            const xAxis = ((e.pageY - rect.top) - rect.height / 2) / 20;
            const yAxis = (rect.width / 2 - (e.pageX - rect.left)) / 20;
            card.style.transform
                = `perspective(1000px) rotateX(${xAxis}deg) rotateY(${yAxis}deg)`;
        });

        card.addEventListener('mouseleave', function (e) {
            clearTimeout(timeoutID);
            card.style.transition = 'all 0.5s ease';
            card.style.transform = `perspective(1000px) rotateX(0deg) rotateY(0deg)`;
        });
    }

    initialize() {
        const moon = document.getElementById('moon');
        const sun = document.getElementById('sun');

        const prefersDarkScheme = window
            .matchMedia("(prefers-color-scheme: dark)").matches;
        this.setTheme(prefersDarkScheme ? 'dark' : 'light');
        if (prefersDarkScheme) {
            moon.style.display = 'none';
            sun.style.display = 'block';
        } else {
            moon.style.display = 'block';
            sun.style.display = 'none';
        }
        this.populateProfile();
        this.populatePersonalLinks();
        this.populateSocialLinks();
        this.initializeThemeToggle(moon, sun);
        this.initializeCardTransform();
    }

}

const profileCard = new ProfileCard(
    profileCardConfig.profile,
    profileCardConfig.personalLinks,
    profileCardConfig.socialLinks,
    profileCardConfig.themes
);

document.addEventListener("DOMContentLoaded", function () {
    profileCard.initialize();
    document.body.style.display = ""; // Remove the inline style
});


config.js
const profileCardConfig = {
    profile: {
        picture: "http://placekitten.com/g/200/300",
        headline: "Kitten",
        description: "Your description goes here...Your description goes here...",
    },
    personalLinks: [
        { text: "Portfolio", url: "#" },
        { text: "Contact", url: "#" },
    ],
    socialLinks: [
        { icon: "logo-github", url: "#" },
        { icon: "logo-mastodon", url: "#" },
        { icon: "logo-twitter", url: "#" },
        { icon: "logo-instagram", url: "#" },
    ],
    themes: {
        light: {
            "--bg-color": "#fff",
            "--text-color": "#000",
            "--card-bg-color": "#e4e4e4",
            "--accent-color": "#007BFF",
        },
        dark: {
            "--bg-color": "#1a1a1a",
            "--text-color": "#fff",
            "--card-bg-color": "#333",
            "--accent-color": "#00BCD4",
        },
    },
};

export { profileCardConfig };